pkg/topology/full/full.go

```diff
@@ -69,11 +70,13 @@ func (d *Driver) AddPeer(ctx context.Context, addr swarm.Address) error {
 	if !isConnected(addr, connectedPeers) {
 		peerAddr, err := d.p2pService.Connect(ctx, ma)
 		if err != nil {
-			return err
+			if !errors.Is(err, p2p.ErrAlreadyConnected) {
+				return err
+			}
 		}
 
 		// update addr if it is wrong or it has been changed
-		if !addr.Equal(peerAddr) {
+		if !addr.Equal(peerAddr) && !peerAddr.IsZero() {
```

Comment:
why do we need this `IsZero` check? why should zero addresses be transmitted in the first place?

Reply:
Few lines aboce I check `IsConnected` which is not thread-safe, and return an error only if it's not `ErrAlreadyConnected`. If it is, the code continues, and `peerAddr` will be zero in this case.

Reply:
i'm sorry, i might have missed something, but why should the code continue execution when there is an error in the code block above. can you elaborate? why would you like to put the peer again in to the addressbook if its already there?

Reply:
Well, basically, if you see, the code logic is that it connects if the node is not already connected. So, I don't see a reason to break it on already connected error, cuz it would be the same thing, right?

Reply:
Actually, you are right, it is the different error actually, so I will break in this case as well. It will make the code more simple, and it is still good enough. Thanks!